infrastructure/src/main/java/org/corfudb/infrastructure/BaseServer.java

```diff
@@ -78,49 +112,54 @@ private void keepAlive(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r)
         r.sendResponse(ctx, msg, CorfuMsgType.ACK.msg());
     }
 
+    //TODO(Zach): Do we need a KEEP_ALIVE?
+
     /**
      * Respond to a version request message.
      *
-     * @param msg   The incoming message
+     * @param req   The incoming message
      * @param ctx   The channel context
      * @param r     The server router.
      */
-    @ServerHandler(type = CorfuMsgType.VERSION_REQUEST)
-    private void getVersion(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
-        VersionInfo vi = new VersionInfo(serverContext.getServerConfig(),
-                                         serverContext.getNodeIdBase64());
-        r.sendResponse(ctx, msg, new JSONPayloadMsg<>(vi, CorfuMsgType.VERSION_RESPONSE));
+    @RequestHandler(type = RequestPayloadMsg.PayloadCase.VERSION_REQUEST)
+    private void getVersion(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        VersionInfo vi = new VersionInfo(serverContext.getServerConfig(), serverContext.getNodeIdBase64());
+        HeaderMsg responseHeader = getHeaderMsg(req.getHeader(), false, true);
+        ResponseMsg response = getResponseMsg(responseHeader, getVersionResponseMsg(vi));
+        r.sendResponse(response, ctx);
     }
 
     /**
-     * Respond to a epoch change message.
+     * Respond to a epoch change request.
      * This method also executes sealing logic on each individual server type.
      *
-     * @param msg The incoming message
-     * @param ctx The channel context
-     * @param r   The server router.
+     * @param req The incoming request message.
+     * @param ctx The channel context.
+     * @param r The server router.
      */
-    @ServerHandler(type = CorfuMsgType.SEAL)
-    public synchronized void handleMessageSetEpoch(@NonNull CorfuPayloadMsg<Long> msg,
-                                                   ChannelHandlerContext ctx,
-                                                   @NonNull IServerRouter r) {
+    @RequestHandler(type = RequestPayloadMsg.PayloadCase.SEAL_REQUEST)
+    private synchronized void handleSeal(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
         try {
-            long epoch = msg.getPayload();
+            final long epoch = req.getPayload().getSealRequest().getEpoch();
             String remoteHostAddress;
             try {
                 remoteHostAddress = ((InetSocketAddress)ctx.channel().remoteAddress()).getAddress().getHostAddress();
-            } catch (NullPointerException e) {
+            } catch(NullPointerException ex) {
```

Comment:
I know you haven't written this code, but how can the remoteAddress be unavailable if this call is invoked after the tcp connection is established?

Reply:
Also, a cast to `InetSocketAddress` is not required because implementations of `SocketAddress` override the `toString` method.

Reply:
Is it possible that the connection becomes unavailable between this method's entry point and the call to remoteAddress?

Reply:
Yes, it is possible, but why would it unset the SocketAddress if it becomes unavailable? 

There are other APIs to determine availability (i.e., https://netty.io/4.0/api/io/netty/channel/Channel.html#isOpen--)